File: 11.container-with-most-water.cpp
/*
 * @lc app=leetcode id=11 lang=cpp
 *
 * [11] Container With Most Water
 */
// Explanation:
// https://leetcode.com/problems/container-with-most-water/solutions/2363527/the-most-clear-explanation-of-the-greedy-approach/?q=greedy+explan&orderBy=most_relevant

    //smaller height defines the height of the container. 
    // Moving smaller border may lead to a bigger container
    // Moving greater border always leads to a smaller container

// @lc code=start
class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0;
        int right = height.size()-1;
        int maxArea = 0;
        while (left < right) {
            maxArea = max(maxArea, (right-left) * min(height[left], height[right]));
            if (height[left] < height[right]) {
                left++;
            }
            else {
                right--;
            }
        }
        return maxArea;
    }
};
// @lc code=end

File: 11.container-with-most-water.py
#
# @lc app=leetcode id=11 lang=python
#
# [11] Container With Most Water
#

# @lc code=start
class Solution(object):
    def maxArea(self, height):
        l = 0
        r = len(height) - 1
        maxVolume = 0
        while (l < r):
            volume = (r - l) * min(height[l], height[r])
            maxVolume = max(volume, maxVolume)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return maxVolume
                
    # def maxArea(self, height):
    #     """
    #     :type height: List[int]
    #     :rtype: int
    #     """
    #     i = 0
    #     j = len(height) - 1
    #     max_vol = 0
    #     while i < j:
    #         shorter_bar_index = i if height[i] < height[j] else j
    #         curr_vol = (j - i) * height[shorter_bar_index]
    #         max_vol = curr_vol if curr_vol > max_vol else max_vol
    #         if height[i] < height[j]:
    #             i += 1
    #         else:
    #             j -= 1
    #     return max_vol


# @lc code=end
